### PR TITLE
Load internal json v3

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This is an early PR to start loading the internal JSON v3 representation added in https://github.com/Homebrew/brew/pull/16541. The hope is that this will not only load faster than the previous one but also will reduce the amount of data that has to flow across the network and reduce the amount of code needed to turn the current JSON representation into the expected shape internally.

My first thought is that the amount of code needed to get this to work is way less than I originally thought. Of course, no tests so far so this will grow a bit.

- https://github.com/Homebrew/brew/issues/16410

### CODE REVIEW

**Comment by apainintheneck** [comment]
Early benchmarks on my super old iMac show that this is noticeably faster. I'll try it on a newer machine at some point to get more relevant information. Note that `HOMEBREW_INTERNAL_JSON_V3` just gets ignored on the master branch.

```console
$ hyperfine --parameter-list branch master,load-internal-json-v3 --warmup 3 --setup 'git switch {branch}' 'HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl'
Benchmark 1: HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = master)
  Time (mean ± σ):      3.164 s ±  0.020 s    [User: 2.086 s, System: 0.915 s]
  Range (min … max):    3.133 s …  3.196 s    10 runs

Benchmark 2: HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = load-internal-json-v3)
  Time (mean ± σ):      2.731 s ±  0.020 s    [User: 1.694 s, System: 0.877 s]
  Range (min … max):    2.710 s …  2.783 s    10 runs

Summary
  HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = load-internal-json-v3) ran
    1.16 ± 0.01 times faster than HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = master)
```

**Comment by MikeMcQuaid** [comment]
> Early benchmarks on my super old iMac show that this is noticeably faster. I'll try it on a newer machine at some point to get more relevant information. Note that `HOMEBREW_INTERNAL_JSON_V3` just gets ignored on the master branch.

Wow, that's huge! Great work @apainintheneck!

File: Library/Homebrew/extend/cachable.rb [hunk 1]
```diff
@@ -7,6 +7,11 @@ def cache
     @cache ||= T.let({}, T.nilable(T::Hash[T.untyped, T.untyped]))
   end
 
+  sig { params(hash: T.nilable(T::Hash[T.untyped, T.untyped])).void }
+  def overwrite_cache!(hash)
+    @cache = hash
+  end
+
```
Comment (Bo98) [inline Library/Homebrew/extend/cachable.rb:14]: Ideally this wouldn't be public otherwise you could do `Formulary.overwrite_cache! {}`

Comment (apainintheneck) [reply Library/Homebrew/extend/cachable.rb:14]: Good point!

**Comment by apainintheneck** [comment]
Okay, I finally got around to adding some tests. There are tests for both generating the JSON format and loading formulae from the JSON file. I ended up creating a mini-tap fixture to facilitate testing as well. I assume grabbing a few random formulae for that is not a big deal. Let me know if there's anything else that would be good to test for.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      # NOTE: Symlinks can't be copied recursively so we create them manually here.
+      (Tap::TAP_DIRECTORY/"homebrew/homebrew-core").tap do |core_tap|
+        mkdir(core_tap/"Aliases")
+        ln_s(core_tap/"Formula/f/fennel.rb", core_tap/"Aliases/fennel-lang")
+        ln_s(core_tap/"Formula/p/ponyc.rb", core_tap/"Aliases/ponyc-lang")
+      end
+    end
+
+    it "creates the expected hash" do
+      api_hash = CoreTap.instance.to_api_hash
```
Comment (apainintheneck) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:20]: It looks like this line is picking up some unrelated formulae at test time. I guess that means that the test tap is not isolated from test to test so I might have to change my approach here.

My first approach was trying to mock the core tap directory so that I wouldn't have to move and write the files in the first place but I couldn't get the paths to work so I fell back on this approach.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+describe "Internal Tap JSON -- Formula" do
+  let(:internal_tap_json) { File.read(TEST_FIXTURE_DIR/"internal_tap_json/homebrew-core.json").chomp }
+  let(:tap_git_head) { "9977471165641744a829d3e494fa563407503297" }
+
+  context "when generating JSON", :needs_macos do
+    before do
+      cp_r(TEST_FIXTURE_DIR/"internal_tap_json/homebrew-core", Tap::TAP_DIRECTORY/"homebrew")
```
Comment (apainintheneck) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Sometimes it seems like some of these files don't get copied over or get deleted before the tests can run. I'm not sure why that is.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:20]:
```
Failures:

  1) Internal Tap JSON -- Formula when generating JSON creates the expected hash
     Failure/Error: api_hash = CoreTap.instance.to_api_hash

     FormulaUnavailableError:
       No available formula with the name "testball".
     # ./formulary.rb:669:in `get_formula'
     # ./formulary.rb:758:in `factory'
     # ./tap.rb:1182:in `block in to_api_hash'
     # ./tap.rb:1181:in `to_h'
     # ./tap.rb:1181:in `to_api_hash'
     # ./test/api/internal_tap_json/formula_spec.rb:20:in `block (3 levels) in <top (required)>'
     # ./test/support/helper/spec/shared_context/homebrew_cask.rb:64:in `block (2 levels) in <top (required)>'
```

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: @apainintheneck is that only when running in parallel or can you reproduce it serially?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
+# frozen_string_literal: true
+
+describe "Internal Tap JSON -- Formula" do
+  let(:internal_tap_json) { File.read(TEST_FIXTURE_DIR/"internal_tap_json/homebrew-core.json").chomp }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:4]: Would prefer to have this specified inline. I know it's relatively large but it'll help us review the actual generated contents more than a single line fixture.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+          "desc"             => "Lua Lisp Language",
+          "full_name"        => "fennel",
+          "homepage"         => "https://fennel-lang.org",
+          "installed"        => [],
+          "license"          => "MIT",
+          "name"             => "fennel",
+          "ruby_source_path" => "Formula/f/fennel.rb",
+          "tap"              => "homebrew/core",
+          "tap_git_head"     => tap_git_head,
+          "versions"         => { "bottle"=>true, "head"=>nil, "stable"=>"1.4.0" },
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:82]:
```suggestion
          "dependencies"     => ["lua"],
          "desc"             => "Lua Lisp Language",
          "full_name"        => "fennel",
          "homepage"         => "https://fennel-lang.org",
          "license"          => "MIT",
          "name"             => "fennel",
          "ruby_source_path" => "Formula/f/fennel.rb",
          "tap"              => "homebrew/core",
          "tap_git_head"     => tap_git_head,
          "versions"         => { "bottle"=>true, "stable"=>"1.4.0" },
```

As these don't seem to be needed.

Do we use `desc` and `license` from the JSON?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+          "desc"                   => "Object-oriented, actor-model, capabilities-secure programming language",
+          "full_name"              => "ponyc",
+          "homepage"               => "https://www.ponylang.io/",
+          "license"                => "BSD-2-Clause",
+          "name"                   => "ponyc",
+          "ruby_source_path"       => "Formula/p/ponyc.rb",
+          "tap"                    => "homebrew/core",
+          "tap_git_head"           => tap_git_head,
+          "uses_from_macos"        => [{ "llvm"=>[:build, :test] }, "zlib"],
+          "uses_from_macos_bounds" => [{}, {}],
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:97]: Sorry to increase up scope here potentially (and can be in another PR) but: this API is fairly awful due to backwards compatibility concerns we don't have for v3 so would be nice to address. CC @Bo98 for thoughts on what that should look like.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+          "versions"               => { "bottle"=>true, "head"=>nil, "stable"=>"0.58.1" },
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:98]:
```suggestion
          "versions"               => { "bottle"=>true, "stable"=>"0.58.1" },
```

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
+class Fennel < Formula
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Might be nicer to strip these down to only the bits that are actually used in the API (e.g. less in `def install`, `test do`)

**Review by MikeMcQuaid — COMMENTED** [review]
Great work @apainintheneck 👏🏻

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (Bo98) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:97]: Yeah we were going to cycle back and clean this up. I can open a PR with suggestions

Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:97]: Cool: given that: fine to merge with this as-is in this PR 👍🏻. A TODO or something might be nice though?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: It doesn't fail when running one file but does when running the whole test suite. I'll try running everything serially to see what happens.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 4 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:4]: You can just go look at the file though, right?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 5 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:82]: Yeah, I'll remove the installed array. That's just an oversight on my part. The fields here are not super important we're just trying to see that it loads somewhat correctly. It's a sanity check. Description and license are used by `brew info` and description is also used by `brew search` and `brew desc`.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:98]: Yeah, I can probably compact or avoid adding that field in the first place.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb (hunk 8 above)
Comment (apainintheneck) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Good point.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Everything passes when running the tests serially locally.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb (hunk 8 above)
Comment (apainintheneck) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Apparently removing them causes the tests to fail for some reason. 🤷

Since it's not that important, I'll just leave them as is.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:98]: Never mind, this only exists in the public facing JSON and should already be omitted from the internal one. I'm just using the public JSON here for testing because it exercises more parts of the formula code.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 4 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:4]: @apainintheneck Technically, yeh, but then I need to mentally convert a JSON string to a Ruby hash. If it was at least pretty-printed in that file (indented, line breaks), it'd make it less painful to e.g. drop comments on that file where stuff shouldn't be present.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Yeh, smells like this is just multiple tests creating/deleting the same files at the same time. May need to have a randomised/test runner based filename/directory name.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:97]: Yeh, I think it'd be preferable to use the internal JSON and to get full coverage of the formula code through the input JSON.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb (hunk 8 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Sorry, I don't mean remove those blocks, poor communication on my part, but instead make the contents of each e.g. a one-liner/comment.

Comment (apainintheneck) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Yeah, that's what I tried and the tests started failing at that point. Not obvious to me what difference it makes though and the diff output seemed pretty much identical.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core/Formula/f/fennel.rb:1]: Ok, no worries, ignore it!

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Interesting... my mental model of how the tests work is that each process would have a separate temporary directory and wouldn't pollute each others workspace. The test directories also seem to get cleaned up after each test.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:97]: > Yeh, I think it'd be preferable to use the internal JSON and to get full coverage of the formula code through the input JSON.

Sorry, I don't know what you mean by this. How can we get full coverage of the formula code through the input JSON? What input JSON are we referring to here?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Only if you're creating a new e.g. tmpdir for each. Otherwise: use same filename in multiple tests, run them in parallel and they will happily cleanup the WIP of another test.

Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: That seems like a major bug in the test suite then, doesn't it? We should be able to create one temp directory per process and that would allow for better reproducibility in that case.

File: Library/Homebrew/test/.irb_history
Comment (apainintheneck) [inline Library/Homebrew/test/.irb_history:1]: Should be ignored. I'll update `.gitignore`.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: It looks like each process gets a temporary directory already which is very good. The code that makes that work is here.

https://github.com/Homebrew/brew/blob/c35546167da22c3c00d34332ff84a1bb40153766/Library/Homebrew/test/support/lib/startup/config.rb#L8-L16

Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:9]: Yeh, there's definitely some temporary handling but there's also definitely ways even with that that you can inadvertently add/remove files in different tests running in parallel (this has bitten me before). Good luck!

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 4 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:4]: (some of the constants change between tests/test runners and others stay the same. some tests likely are modifying shared state when they shouldn't)

**Comment by apainintheneck** [comment]
This should be ready for review again. The test failures were caused by caching problems between test runs. I just needed to clear the caches a bit more between them and add one more variable to the `Tap.clear_cache` method. https://github.com/Homebrew/brew/pull/16710 solved some of those problems though which was helpful. https://github.com/Homebrew/brew/pull/16671 could help solve these sorts of problems more in the future.

I'm leaving this as draft until after the Monday release since I'd like to be cautious and a have a few days before it's available to everyone to avoid any unexpected bugs.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json [hunk 9]
```diff
@@ -0,0 +1 @@
...350"}},"post_install_defined":false,"ruby_source_path":"Formula/p/ponyc.rb","ruby_source_sha256":"81d51c25d18710191beb62f9f380bae3d878aad815a65ec1ee2a3b132c1fadb3","build_dependencies":["cmake","python@3.12"],"uses_from_macos":[{"llvm":["build","test"]},"zlib"],"uses_from_macos_bounds":[{},{}],"version":"0.58.1","bottle":{"files":{"arm64_sonoma":{"cellar":":any_skip_relocation","sha256":"e3aecfcf02aea56d53d82691e2ad7a780f771023d7070271bfce96b17439a34d"},"arm64_ventura":{"cellar":":any_skip_relocation","sha256":"6ff83717191e16e4f852fb3be8f838afba312cc39e601bb5cebd2a618a328658"},"arm64_monterey":{"cellar":":any_skip_relocation","sha256":"25c91bce200583a96f4cea34f31393c8f10eadcab363cc7d4d864d15f5f97e25"},"sonoma":{"cellar":":any_skip_relocation","sha256":"5f4c550ce33e2970e0ada18a409755fa62936181289a21c15582ff80343866b6"},"ventura":{"cellar":":any_skip_relocation","sha256":"f26c799f45013685da779bf2008ebe1907f9b3a93d5f260ce271a3f3b628da50"},"monterey":{"cellar":":any_skip_relocation","sha256":"1cff10d068b36b18b253d235424c4f5aef71ff9ee44f2522c4b041dd4383ec30"},"x86_64_linux":{"cellar":":any_skip_relocation","sha256":"ab49318d75eed3ee932c8e5add22f252ec0c852aad94945022877f926e93899f"}}}}}}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json:1]: Can this be pretty printed please so it's easier to review?

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]: Can a TODO be added here please.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: I'm saying it seems very weird to be to use the non-internal public JSON for testing the internal JSON format?

Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: It's not that weird. It essentially mimics `brew readall`.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]: I left a not in the original issue thread: https://github.com/Homebrew/brew/issues/16410#issuecomment-1963079892

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json (hunk 9 above)
Comment (apainintheneck) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json]: Addressed in https://github.com/Homebrew/brew/pull/16638/commits/4a8b20125fa5223cda028da7067f6e5a3a87d4bb

**Comment by apainintheneck** [comment]
Here are the same benchmarks on my M1 MacBook that I use at work. Still significant performance boost though not as big as the one on my older computer. Percentage-wise it's more significant though.

```console
$ hyperfine --parameter-list branch master,load-internal-json-v3 --warmup 3 --setup 'git switch {branch}' 'HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl'
Benchmark 1: HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = master)
  Time (mean ± σ):      1.079 s ±  0.010 s    [User: 0.738 s, System: 0.164 s]
  Range (min … max):    1.070 s …  1.096 s    10 runs

Benchmark 2: HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = load-internal-json-v3)
  Time (mean ± σ):     876.8 ms ±  10.4 ms    [User: 554.8 ms, System: 146.4 ms]
  Range (min … max):   867.7 ms … 895.6 ms    10 runs

Summary
  HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = load-internal-json-v3) ran
    1.23 ± 0.02 times faster than HOMEBREW_INTERNAL_JSON_V3=1 brew info jrnl (branch = master)
```

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: @apainintheneck Ok but aren't these JSON formats going to diverge significantly? 

I think I might just need an ELI5 treatment here, sorry.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]: @apainintheneck I'd really love an inline Ruby comment here if possible.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      ],
+      "uses_from_macos_bounds": [
+        {
+        },
+        {
+        }
+      ],
+      "version": "0.58.1",
+      "bottle": {
+        "files": {
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json:83]: Could maybe strip this key

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks @apainintheneck. A couple of small requests but I'm fine if this is merged as-is and they are addressed in follow-up PRs.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (Bo98) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: This is basically a fancy `to_hash` test. We could do it as a series of checks instead, though would be multiple `expects` in a single check so might be long:

```rb
expect(formula.name).to eq("something")
expect(formula.version.to eq("1.0")
expect(formula.revision to be_zero)
# ...
```

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]: It seems weird to leave a comment when any potential changes are exploratory. We haven't decided on the form that we want yet.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 7 above)
Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: `#to_hash` is a bit better than doing a bunch of individual checks since it will end up calling more methods internally to fill out the fields we're not checking.

The goal here is not to validate the `#to_hash` method but instead to be reasonably confident that the formula loads without failing. `#to_hash` is just an easy way to take a peek inside the formula instance and run some arbitrary methods along the way. I'm open to alternatives that would be clearer, briefer or provide more assurance of correctness. The current test is not comprehensive at all.

`#to_api_hash` is already being exercised by a different test where we generate the JSON API for the core formula tap.

**Comment by apainintheneck** [comment]
I'll merge this in tomorrow.

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]: @apainintheneck It's a hard `TODO` because we're not going to ship the internal JSON v3 without changing this. I'm fine on not blocking the PR on this stuff but I would really like something beyond a comment in another issue to track this in the actual code.

File: Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json (hunk 10 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/support/fixtures/internal_tap_json/homebrew-core.json:83]: Ok, thanks, that's fine then 👍🏻

File: Library/Homebrew/test/api/internal_tap_json/formula_spec.rb (hunk 6 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:112]:
```suggestion
          # TODO: improve this API before we ship internal API v3 to users
          "uses_from_macos"        => [{ "llvm"=>[:build, :test] }, "zlib"],
          "uses_from_macos_bounds" => [{}, {}],
```

Comment (apainintheneck) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: > @apainintheneck It's a hard `TODO` because we're not going to ship the internal JSON v3 without changing this. I'm fine on not blocking the PR on this stuff but I would really like something beyond a comment in another issue to track this in the actual code.

I'm fine with the comment but this is news to me. I didn't know that this was that important either way.

Comment (MikeMcQuaid) [reply Library/Homebrew/test/api/internal_tap_json/formula_spec.rb:113]: Apologies for my unclear communication! I'm trying to get the balance right between being overly pedantic/demotivating and blocking individual PRs vs. making clear what's required before we ship to users. Great work getting this shipped, thanks @apainintheneck ❤️
